src/riak_cs_storage_mr.erl

```diff
@@ -0,0 +1,280 @@
+-spec new_or_old(erlang:timestamp(), erlang:timestamp()) -> new | old.
+new_or_old(LeewayEdge, {_,_,_} = Timestamp) when LeewayEdge < Timestamp -> new;
+new_or_old(_, _) -> old.
+
+-spec add_to(sum(), pos_integer(),
+             {non_neg_integer(), non_neg_integer(), non_neg_integer()}) -> sum().
+add_to(Sum, Pos, {Count, Bytes, Blocks}) ->
+    {C0, By0, Bl0} = element(Pos, Sum),
+    setelement(Pos, Sum, {C0 + Count, By0 + Bytes, Bl0 + Blocks}).
+
+%% @doc Convert `sum()' record to list.
+-spec summary_to_list(sum()) -> [{term(), non_neg_integer()}].
+summary_to_list(Sum) ->
+    [_RecName | Triples] = tuple_to_list(Sum),
+    summary_to_list(record_info(fields, sum), Triples, []).
+
+summary_to_list([], _, Acc) ->
+    Acc;
+summary_to_list([F|Fields], [{C, By, Bl}|Triples], Acc) ->
+    summary_to_list(Fields, Triples,
+                    [{{F, objects}, C}, {{F, bytes}, By}, {{F, blocks}, Bl} | Acc]).
+
+object_size(Resolved) ->
+    {MPparts, MPbytes} = count_multipart_parts(Resolved),
+    case riak_cs_manifest_utils:active_manifest(Resolved) of
+        {ok, ?MANIFEST{content_length=Length}} ->
+            [{1 + MPparts, Length + MPbytes}];
+        _ ->
+            [{MPparts, MPbytes}]
+    end.
+
+-spec count_multipart_parts([{cs_uuid(), lfs_manifest()}]) ->
+                                   {non_neg_integer(), non_neg_integer()}.
+count_multipart_parts(Resolved) ->
+    lists:foldl(fun count_multipart_parts/2, {0, 0}, Resolved).
+
+-spec count_multipart_parts({cs_uuid(), lfs_manifest()},
+                            {non_neg_integer(), non_neg_integer()}) ->
+                                   {non_neg_integer(), non_neg_integer()}.
+count_multipart_parts({_UUID, ?MANIFEST{props=Props, state=writing} = M},
+                      {MPparts, MPbytes} = Acc)
+  when is_list(Props) ->
+    case proplists:get_value(multipart, Props) of
+        ?MULTIPART_MANIFEST{parts=Ps} = _  ->
+            {MPparts + length(Ps),
+             MPbytes + lists:sum([P?PART_MANIFEST.content_length ||
+                                     P <- Ps])};
+        undefined ->
+            %% Maybe not a multipart
+            Acc;
+        Other ->
+            %% strange thing happened
+            _ = lager:log(warning, self(),
+                          "strange writing multipart manifest detected at ~p: ~p",
+                          [M?MANIFEST.bkey, Other]),
+            Acc
+    end;
+count_multipart_parts(_, Acc) ->
+    %% Other state than writing, won't be counted
+    %% active manifests will be counted later
+    Acc.
+
+bucket_summary_reduce(Sums, _) ->
+    InitialCounters = orddict:new(),
+    [bucket_summary_fold(Sums, InitialCounters)].
+
+bucket_summary_fold([], Counters) ->
+    Counters;
+bucket_summary_fold([Sum | Sums], Counters) ->
+    NewCounters = lists:foldl(
+                    fun({K, Num}, Cs) -> orddict:update_counter(K, Num, Cs) end,
+                    Counters, Sum),
+    bucket_summary_fold(Sums, NewCounters).
```

Comment:
Maybe this might be simpler?
```erlang
bucket_summary_fold(Sums) ->
   lists:foldl(fun(Sum, Acc) -> orddict:merge(fun(K, V1, V2) -> V1+V2 end, Sum, Acc) end,
                     orddict:new(), Sums).
```